workflows/prognostic_c48_run/runtime/steppers/reservoir.py

```diff
@@ -0,0 +1,233 @@
+import pandas as pd
+from typing import Optional, MutableMapping, Hashable, Sequence, Union, cast
+import xarray as xr
+
+import fv3fit
+from fv3fit._shared import put_dir
+from fv3fit._shared.halos import append_halos_using_mpi
+from fv3fit.reservoir.model import HybridReservoirComputingModel
+from fv3fit.reservoir.adapters import ReservoirDatasetAdapter
+
+
+ReservoirAdapter = Union[ReservoirDatasetAdapter, HybridReservoirComputingModel]
+
+
+@dataclasses.dataclass
+class ReservoirConfig:
+    """
+    Reservoir model configuration.
+
+    Attributes:
+        model: URL to the global hybrid RC model
+    """
+
+    models: Sequence[str]
+    synchronize_steps: int = 1
+    reservoir_timestep: str = "3h"  # TODO: Could this be inferred?
+
+
+class _FiniteStateMachine:
+
+    INCREMENT = "increment"
+    PREDICT = "predict"
+
+    def __init__(self, init_time: Optional[cftime.DatetimeJulian] = None) -> None:
+        self._last_called = None
+        self._init_time = init_time
+        self._num_increments_completed = 0
+
+    @property
+    def init_time(self):
+        return self._init_time
+
+    @property
+    def completed_increments(self):
+        return self._num_increments_completed
+
+    def to_incremented(self):
+        # incrementing allowed anytime, e.g., synchronizing
+        self._last_called = self.INCREMENT
+        self._num_increments_completed += 1
+
+    def to_predicted(self):
+        # predict only allowed after increment has been called
+        if self._last_called != self.INCREMENT:
+            raise ValueError("Must call increment before next prediction")
+        self._last_called = self.PREDICT
+
+    def set_init_time(self, time: cftime.DatetimeJulian):
+        self._init_time = time
+
+    def __call__(self, state: str):
+        if state == self.INCREMENT:
+            self.to_incremented()
+        elif state == self.PREDICT:
+            self.to_predicted()
+        else:
+            raise ValueError(
+                f"Unknown state provided to _ReservoirStepperState {state}"
+            )
+
+
+class _ReservoirStepper:
+
+    label = "hybrid_reservoir"
+
+    def __init__(
+        self,
+        model: ReservoirAdapter,
+        reservoir_timestep: timedelta,
+        synchronize_steps: int,
+        model_timestep_seconds: int = 900,
+        state_machine: Optional[_FiniteStateMachine] = None,
+    ):
+        self.model = model
+        self.rc_timestep = reservoir_timestep
+        self.synchronize_steps = synchronize_steps
+        self.dt_atmos = timedelta(seconds=model_timestep_seconds)
+
+        if state_machine is None:
+            self._state_machine = _FiniteStateMachine()
+        else:
+            self._state_machine = state_machine
+
+    @property
+    def init_time(self):
+        return self._state_machine.init_time
+
+    @property
+    def completed_sync_steps(self):
+        return self._state_machine.completed_increments
+
+    def increment_reservoir(self, time, state):
```

Comment:
Is there a reason why this method and `predict` are on the base class rather than their respective child classes?

Reply:
Mainly for ease of testing, but I will co-locate the code with the point of use!